docs/SILProgrammersManual.md

```diff
@@ -159,6 +159,401 @@ rules out PartialApplies is no excuse to conflate applied arguments
 with function arguments. Also, without consistent use of common
 idioms, it becomes overly burdensome to evolve these APIs over time.
 
+## `AccessedStorage` and `AccessPath`
+
+The `AccessedStorage` and `AccessPath` types formalize memory access
+in SIL. Given an address-typed SIL value, it must be possible to
+reliably identify the storage location of the accessed
+memory. `AccessedStorage` identifies an accessed storage
+location. `AccessPath` contains both a storage location and the
+"access path" within that memory object. The relevant API details are
+documented in MemAccessUtils.h
+
+### Formal access
+
+SIL preserves the language semantics of formal variable access in the
+form of access markers. `begin_access` identifies the address of the
+formal access and `end_access` delimits the scope of the access. At
+the language level, a formal access is an access to a local variable
+or class property. For details, see
+[SE-0176: Enforce Exclusive Access to Memory](https://github.com/apple/swift-evolution/blob/master/proposals/0176-enforce-exclusive-access-to-memory.md)
+
+Access markers are preserved in SIL to:
+
+1. verify exclusivity enforcement
+
+2. continue optimizing exclusivity checks following other transforms
+
+3. simplify and strengthen general analyses of memory access. For
+example, `begin_access [read] %address` indicates that the accessed
+address is immutable for the duration of its access scope
+
+### Access path def-use relationship
+
+Computing `AccessedStorage` and `AccessPath` for any given SIL address
+involves a use-def traversal to determine the origin of the
+address. It may traverse operations on address, pointer, box, and
+reference types. The logic that formalizes which SIL operations may be
+involved in the def-use chain is encapsulated with the
+`AccessUseDefChainVisitor`. The traversal can be customized by
+implementing this visitor. Customization is not expected to change the
+meaning of AccessedStorage or AccessPath. Rather, it is intended for
+additional pass-specific book-keeping or for higher-level convenience
+APIs that operate on the use-def chain bypassing AccessedStorage
+completely.
+
+Access def-use chains are divided by three points: the "root", the
+formal access "base", and the address of a memory operation. For
+example:
+```
+  struct S {
+    var field: Int64
+  }
+  class C {
+    var prop: S
+  }
+
+  %root    = alloc_ref $C
+  %base    = ref_element_addr %root : $C, #C.prop
+  %access  = begin_access [read] [static] %base : $*S
+  %memaddr = struct_element_addr %access : $*S, #.field
+  %value   = load [trivial] %memaddr : $*Int64
+  end_access %access : $*S
+```
+
+#### Reference root
+
+The first part of the def-use chain computes the formal access base
+from the root of the object (e.g. `alloc_ref ->
+ref_element_addr`). Only reference types or Builtin.BridgeObject types
```

Comment:
So you know, the implementation of reference roots is not in this PR because I want to enable verification one step at a time without breaking too much.

```
The first part of the def-use chain computes the formal access base
from the root of the object (e.g. `alloc_ref ->
ref_element_addr`). The reference root might be a locally allocated
object, a function argument, a function result, or a reference loaded
from storage. There is no enforcement on the type of operation that
can produce a reference; however, only reference types or
Builtin.BridgeObject types are only allowed in this part of the
def-use chain. The reference root is the greatest common ancestor in
the def-use graph that can identify an object by a single SILValue. If
the root as an `alloc_ref`, then it is *uniquely identified*. The
def-use chain from the root to the base may contain reference casts
(`isRCIdentityPreservingCast`) and phis.
```